dcos/scripts/bootstrap.py

```diff
@@ -0,0 +1,190 @@
+
+
+def main():
+    try:
+        jenkins_agent_user = os.environ['JENKINS_AGENT_USER']
+        jenkins_agent_role = os.environ['JENKINS_AGENT_ROLE']
+        jenkins_home_dir = os.environ['JENKINS_HOME']
+        jenkins_framework_name = os.environ['JENKINS_FRAMEWORK_NAME']
+        jenkins_app_context = os.environ['JENKINS_CONTEXT']
+        marathon_nginx_port = os.environ['PORT0']
+        marathon_jenkins_port = os.environ['PORT1']
+        mesos_master = os.environ['JENKINS_MESOS_MASTER']
+        ssh_known_hosts = os.environ['SSH_KNOWN_HOSTS']
+        marathon_name = os.environ['MARATHON_NAME']
+    except KeyError as e:
+        # Since each of the environment variables above are set either in the
+        # DCOS marathon.json or by Marathon itself, the user should never get
+        # to this point.
+        print("ERROR: missing required environment variable {}.".format(e.args[0]))
+        return 1
+
+    # optional environment variables
+    jenkins_root_url = os.getenv(
+        'JENKINS_ROOT_URL',
+        mesos_dns_taskname(jenkins_framework_name, marathon_name, marathon_nginx_port))
+
+    populate_jenkins_config_xml(
+        os.path.join(jenkins_home_dir, 'config.xml'),
+        mesos_master,
+        jenkins_framework_name,
+        marathon_nginx_port,
+        jenkins_agent_role,
+        jenkins_agent_user,
+        marathon_name)
+
+    populate_jenkins_location_config(os.path.join(
+        jenkins_home_dir, 'jenkins.model.JenkinsLocationConfiguration.xml'),
+        jenkins_root_url)
+
+    populate_known_hosts(ssh_known_hosts, '/etc/ssh/ssh_known_hosts')
+
+    # nginx changes here are really "run once". The context should never
+    # change as long as a Jenkins instance is alive, since the rewrite will
+    # be based on the app ID in Marathon, as will the volume on disk.
+    populate_nginx_config(
+        '/etc/nginx/nginx.conf',
+        marathon_nginx_port,
+        marathon_jenkins_port,
+        jenkins_app_context)
+
+
+def _get_xml_root(config_xml):
+    """Return the ET tree and root XML element.
+
+    :param config_xml: path to config XML file
+    :type config_xml: str
+    :return: a tuple (tree,root)
+    :rtype: tuple
+    """
+    tree = ET.parse(config_xml)
+    root = tree.getroot()
+    return tuple([tree, root])
+
+
+def _find_and_set(element, term, new_text, write_if_empty=False):
+    """Find the desired term within the XML element and replace
+    its text with text.
+
+    :param element: XML element
+    :type element: xml.etree.ElementTree.Element
+    :param term: XML element to find
+    :type term: str
+    :param new_text: New element text
+    :type new_text: str
+    :param write_if_empty : If set to True, the value is updated only if empty.
+                            If set to False, the value is always updated.
+    :type write_if_empty: bool
+    """
+    if not write_if_empty or write_if_empty and not element.find(term).text:
```

Comment:
I really dislike that we mutate a stateful config file like this. Can we please at least add an xml comment to make it clear that this config file was modified by the bootstrap script?

Reply:
We should use https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/README.md for the Jenkins configuration anyways. However, do we want to configure it on each restart? The Jenkins port changes each time it's restarted. Thus the config would have to change. This is a an unstable design. @kvish and @kaiwalyajoshi, what do you think?